### PR TITLE
planner: support more pattern for index join inner side

### LINKED ISSUE #40505: inl_join(a,b) hint doesn't work in update statement
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
--1.  create 2 tables for demo
use test;
create table ods_txn_t_d03(
    id bigint not null auto_increment,
    pnbrn_cnaps varchar(5) not null,
    inst varchar(3) not null,
    txn_accno varchar(18) not null,
    txn_dt date not null,
    ori_rcrd_no varchar(9) not null,
    rcrd_no varchar(9) not null,
    acc_blng_pnbrn_orgno varchar(5) default null,
    msac_accno varchar(18) default null,
    txn_accno_sbacc_type_no varchar(4) default null,
    pd_lgcls varchar(4) default null,
    pd_des varchar(100) default null,
    yn_frz varchar(1) default null,
    txn_curr_tp varchar(3) default null,
    key idx_ods_txn_d_1(id),
    unique key ux_1(pnbrn_cnaps,inst,txn_accno,txn_dt,ori_rcrd_no,rcrd_no)
  );

create table cdm_dic_t(
    pnbrn_cnaps varchar(5) not null,
    new_accno varchar(18) not null,
    primary key(pnbrn_cnaps,new_accno) nonclustered
    );

--2.  use explain to check the plan 
explain 
update 
/*+ use_index(b,), inl_join(b,a) */
    ods_txn_t_d03 b,
    (
        select t1.pnbrn_cnaps,
               t1.new_accno
        from cdm_dic_t t1
        where t1.pnbrn_cnaps = '40001'
     ) a
     set b.yn_frz = '1'
where b._tidb_rowid between 1 and 10000
and b.txn_dt = str_to_date('20221201', '%Y%m%d')
and b.pnbrn_cnaps = a.pnbrn_cnaps
and b.txn_accno = a.new_accno
and b.pnbrn_cnaps = '40001'; 
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
table a and b use IndexJoin
### 3. What did you see instead (Required)
mysql> show warnings;
+---------+------+--------------------------------------------------------------------------------+
| Level   | Code | Message                                                                        |
+---------+------+--------------------------------------------------------------------------------+
| Warning | 1815 | Optimizer Hint /*+ INL_JOIN(b, a) */ or /*+ TIDB_INLJ(b, a) */ is inapplicable |
+---------+------+--------------------------------------------------------------------------------+
### 4. What is your TiDB version? (Required)
mysql> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                     |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v6.1.2
Edition: Community
Git Commit Hash: 5d2030e1d19629b71811c3f14514bad7ed63261a
Git Branch: heads/refs/tags/v6.1.2
UTC Build Time: 2022-10-18 07:57:53
GoVersion: go1.19
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref #40505

Problem Summary:

### What is changed and how it works?

support more pattern for index join inner side 

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by winoros — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -1344,6 +1344,10 @@ type SessionVars struct {
 	// PessimisticTransactionAggressiveLocking controls whether aggressive locking for pessimistic transaction
 	// is enabled.
 	PessimisticTransactionAggressiveLocking bool
+
+	// EnableIndexJoinInnerSideMultiPattern indicates whether enable multi pattern for index join inner side
+	// For now it is not public to user
+	EnableIndexJoinInnerSideMultiPattern bool
```
Comment (qw4990) [inline sessionctx/variable/session.go:1350]: No any SQL variable for this?

Comment (Yisaer) [reply sessionctx/variable/session.go:1350]: right

**Pull request merged**